integration/index_test.go

```diff
@@ -0,0 +1,178 @@
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package integration
+
+import (
+	"encoding/json"
+	"strings"
+	"testing"
+
+	"github.com/awslabs/soci-snapshotter/soci"
+	"github.com/awslabs/soci-snapshotter/util/dockershell"
+	"github.com/containerd/containerd/platforms"
+)
+
+type testImageIndex struct {
+	imgName         string
+	platform        string
+	imgRef          string
+	sociIndexDigest string
+}
+
+func prepareSociIndices(t *testing.T, sh *dockershell.Shell) []testImageIndex {
+	testImages := []testImageIndex{
+		{
+			imgName:  "ubuntu:latest",
+			platform: "linux/arm64",
+		},
+		{
+			imgName:  "alpine:latest",
+			platform: "linux/amd64",
+		},
+		{
+			imgName:  "nginx:latest",
+			platform: "linux/arm64",
+		},
+		{
+			imgName:  "drupal:latest",
+			platform: "linux/amd64",
+		},
+	}
+
+	for i, img := range testImages {
+		platform := platforms.DefaultSpec()
+		if img.platform != "" {
+			var err error
+			platform, err = platforms.Parse(img.platform)
+			if err != nil {
+				t.Fatalf("could not parse platform: %v", err)
+			}
+		}
+		imgInfo := dockerhub(img.imgName, withPlatform(platform))
+		img.imgRef = imgInfo.ref
+		img.sociIndexDigest = optimizeImage(sh, imgInfo)
+		testImages[i] = img
+	}
+
+	return testImages
+}
+
+func TestSociIndexInfo(t *testing.T) {
+	t.Parallel()
+	sh, done := newSnapshotterBaseShell(t)
+	defer done()
+	rebootContainerd(t, sh, "", "")
+
+	testImages := prepareSociIndices(t, sh)
+
+	var sociIndex soci.Index
+
+	for _, img := range testImages {
+		t.Run(img.imgName, func(t *testing.T) {
+			rawJSON := sh.O("soci", "index", "info", img.sociIndexDigest)
+			if err := json.Unmarshal(rawJSON, &sociIndex); err != nil {
+				t.Fatalf("invalid soci index from digest %s: %v", img.sociIndexDigest, rawJSON)
+			}
+			if sociIndex.MediaType != soci.OCIArtifactManifestMediaType {
+				t.Fatalf("invalid soci index media type: expected: %s, actual: %s", soci.OCIArtifactManifestMediaType, sociIndex.MediaType)
+			}
+			if sociIndex.ArtifactType != soci.SociIndexArtifactType {
+				t.Fatalf("invalid soci index artifact type: expected: %s, actual: %s", soci.SociIndexArtifactType, sociIndex.ArtifactType)
+			}
+		})
+	}
+}
+
+func TestSociIndexLs(t *testing.T) {
+	t.Parallel()
+	sh, done := newSnapshotterBaseShell(t)
+	defer done()
+	rebootContainerd(t, sh, "", "")
+
+	testImages := prepareSociIndices(t, sh)
+
+	existHandlerFull := func(output string, img testImageIndex) bool {
+		// full output should have both img ref and soci index digest
+		return strings.Contains(output, img.imgRef) && strings.Contains(output, img.sociIndexDigest)
+	}
+
+	existHandlerQuiet := func(output string, img testImageIndex) bool {
+		// quiet output should only have "soci_index_digest\n"
+		return strings.Contains(output, img.sociIndexDigest+"\n")
```

Comment:
np: Can we do 'strings.Trim(output, "\n"), img.sociIndexDigest` here?

Reply:
The output might have multiple index digests (digest1\ndigest2\n...\n). Trim makes the last one as `digest` but the rest as `digest\n`. Maybe we can use `strings.HasPrefix(output, img.SociIndexDigest+"\n")` which matches the whole line and also works in multi index output?

Reply:
fixed